homeassistant/components/media_player/denonavr.py

```diff
@@ -165,6 +195,24 @@ def update(self):
         self._frequency = self._receiver.frequency
         self._station = self._receiver.station
 
+        if self._sound_mode_support:
+            try:
+                url = ('http://' + str(self._host) +
+                       '/goform/formMainZone_MainZoneXml.xml')
+                xml_data = urllib.request.urlopen(url)
+            except urllib.error.URLError:
+                return
+            parsed_data = ET.parse(xml_data).getroot()
+            sound_mode_raw = parsed_data.find('selectSurround/value')
+            sound_mode_raw = sound_mode_raw.text.rstrip()
+            try:
+                mode_list = list(self._sound_mode_dict.values())
+                mode_index = mode_list.index(sound_mode_raw.upper())
+                sound_mode = list(self._sound_mode_dict.keys())[mode_index]
+                self._current_sound_mode = sound_mode
+            except ValueError:
+                self._current_sound_mode = sound_mode_raw
```

Comment:
The denonavr library does not support sound mode. I have looked at that before I started doing this code myself in Home Assistant.
I do agree that it would be nicer to have this implemented in the denonavr library.
However I do not know the denonavr library at all and do not know any of the developers.
I looked at their code, but it is not as straightforward because they cover all kinds of exceptions and broke up the simple http requests in their big architecture.
Theirfore I chose to just implement it in HASS.

If anyone has the time and knowledge to implement this in denonavr library please do, you can use my code and I am willing to help. But it looked to complicated for me right now and I do not have the time at the moment.

However this code does work and I have tested it.

I propose we go with this for now and maybe someone has time to later implement this further in the denonavr library to get it a little neater.

Reply:
I think you should try to get this merged to denonavr, that's the normal procedure for homeassistant libs anyway, and I think no HA developer is very eager to merge such directly. Maybe you could ask if the authors of that lib are willing to help you to get it included? I think that would be the best way to handle this.